src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs

```diff
@@ -0,0 +1,180 @@
+            dependencies.Add(new DependencyListEntry(factory.NativeLayout.PlacedSignatureVertex(nameAndSig), "Exact method instantiation entry"));
+
+            return dependencies;
+        }
+
+        public void AddEntryIfEligible(NodeFactory factory, MethodDesc method)
+        {
+            // Check if we already saw this method
+            if (!_visitedMethods.Add(method))
+                return;
+
+            if (!IsMethodEligibleForTracking(method))
+                return;
+
+            _exactMethodInstantiationsList.Add(method);
+        }
+
+        private static bool IsMethodEligibleForTracking(MethodDesc method)
+        {
+            // Runtime determined methods should never show up here.
+            Debug.Assert(!method.IsRuntimeDeterminedExactMethod);
+
+            if (method.IsAbstract)
+                return false;
+
+            if (!method.HasInstantiation)
+                return false;
+
+            // This hashtable is only for method instantiations that don't use generic dictionaries,
+            // so check if the given method is shared before proceeding
+            if (method.IsSharedByGenericInstantiations || method.GetCanonMethodTarget(CanonicalFormKind.Specific) != method)
+                return false;
+
+            return true;
+        }
+
+        private Vertex GetNativeLayoutSignatureForMethodEntry(NodeFactory factory, MethodDesc method)
+        {
+            // Get native layout vertices for the declaring type
+
+            ISymbolNode declaringTypeNode = factory.NecessaryTypeSymbol(method.OwningType);
+            Vertex declaringType = _nativeWriter.GetUnsignedConstant(_externalReferences.GetIndex(declaringTypeNode));
+
+            // Get a vertex sequence for the method instantiation args if any
+
+            VertexSequence arguments = new VertexSequence();
+            foreach (var arg in method.Instantiation)
+            {
+                ISymbolNode argNode = factory.NecessaryTypeSymbol(arg);
+                arguments.Append(_nativeWriter.GetUnsignedConstant(_externalReferences.GetIndex(argNode)));
+            }
+
+            // Get the name and sig of the method.
+
+            NativeLayoutMethodNameAndSignatureVertexNode nameAndSig = factory.NativeLayout.MethodNameAndSignatureVertex(method.GetTypicalMethodDefinition());
+            NativeLayoutPlacedSignatureVertexNode placedNameAndSig = factory.NativeLayout.PlacedSignatureVertex(nameAndSig);
+            Debug.Assert(placedNameAndSig.SavedVertex != null);
+
+            // Note: the method name and signature are stored in the NativeLayoutInfo blob, not in the hashtable we build here.
+            Vertex placedNameAndSigOffsetSig = _nativeWriter.GetOffsetSignature(placedNameAndSig.SavedVertex);
+            return _nativeWriter.GetTuple(declaringType, placedNameAndSigOffsetSig, arguments);
```

Comment:
This should be a node by itself, not just written by the ExactMethodInstantiationsNode logic